### ISSUE TITLE
Object method has arguments does not obey strict function types

### ISSUE THREAD

**Issue by ccorcos** [microsoft/TypeScript#48301]
# Bug Report

### 🔎 Search Terms

strictFunctionTypes, class implements interface, method, union type, function arguments

### 🕗 Version & Regression Information

Version: 4.6.2

- [x] This is the behavior in every version I tried, and I reviewed the FAQ for entries.

> "Bugs" that have existed in TS for a long time are very likely to be FAQs; refer to
>   https://github.com/Microsoft/TypeScript/wiki/FAQ#common-bugs-that-arent-bugs

Hmm. This does seem very related to: https://github.com/microsoft/TypeScript/pull/18654 Unfortunately, I'm not familiar with this jargon: "Methods are always bivariant in their argument, while function properties are contravariant in their argument under strictFunctionTypes"

I feel like this is a pretty big problem though!

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/PTAEFkFMBcAsHsAmoBmoDGBDAdqT71IAHaUAV20UhQEttJEAoaATyMlAA1QBeUAb0YBIFAAoAHgH4AXKGxkAtgCNIAJwCUs+crWMAvo0boANpgDOZ0AEFQNBUWOQFkbNEvdBIiVsUqNA4SEQUAAlCmg7DjVVeFVQJTJSbHhQVnZQaNiAQkDg8QwceI4KKlp6RAA6KsDVGDJVXHyAalAARmEDAyN4bDNSJVluPk8xcR8df08gsABlTGdbCzJIKoqauobQZraOgBp9IA)

### 💻 Code

```ts
// Method f can accept undefined
type X = {
	f(x?: number): number
}

class A implements X {
	f(x: number) {
		// Runtime error but no type error!
		// x can be undefined...
		return x + 1
	}
}

const b: X = {
	f(x: number) {
		// Same issue...
		return x + 1
	},
}
```

### 🙁 Actual behavior

There are no type errors here, but I end up with runtime errors because I might call `b.f()` with no type errors but I'll get a runtime error.

### 🙂 Expected behavior

I would expect runtime errors here because the object do not conform to the type that I specified. And I have strict function types on too!

**Comment by MartinJohns** [comment]
Working as intended. See #48085 for more information.

**Comment by fatcerberus** [comment]
> Methods are always bivariant in their argument

This is indeed the relevant bit.  Under `strictFunctionTypes`, parameter types are normally contravariant with respect to the containing function type, which in layman's terms means that the normal subtyping relationship is reversed for the contravariant piece: `number` is assignable to `string | number`, but `(x: number) => void` is *not* assignable to `(x: string | number) => void` (but the reverse is fine).

Method parameters are *bivariant*, which means:
* `meth(x: number): void` is assignable to `meth(x: string | number): void`
* `meth(x: string | number): void` is also assignable to `meth(x: number): void`

In other words, the subtyping relationship for method parameters relative to the containing type goes in both directions.  This ensures that normal assignment works, but *also* that `Array<number>` is assignable to `Array<string | number>`, where that would normally be prohibited due to the type parameter occurring in input positions (e.g. the parameter of `Array#push`).  This is not fully sound as you've noticed, but it's an intentional tradeoff.

**Comment by fatcerberus** [comment]
@ccorcos You may not be familiar with the jargon, but you do seem to intuitively understand that functions *should* be contravariant, which is further than most people get in my experience 😉 (people tend to expect `(x: string) => void` to be assignable to `(x: unknown) => void`, for example, and get confused when you tell them that's wrong).

**Comment by ccorcos** [comment]
Thanks for the help! I had no idea `method(x: string): string` was different from `method: (x: string) => string`! I thought it was just an ambiguous of syntax!

**Comment by ccorcos** [comment]
Interesting example there with `Array#push`... Makes sense, I suppose.

**Comment by DPirate** [comment]
> Thanks for the help! I had no idea `method(x: string): string` was different from `method: (x: string) => string`! I thought it was just an ambiguous of syntax!

Could you please explain your insight? I do not understand what the difference is.

**Comment by MartinJohns** [comment]
@DPirate  
https://www.typescriptlang.org/tsconfig#strictFunctionTypes  
https://github.com/microsoft/TypeScript/wiki/FAQ#why-are-function-parameters-bivariant  
https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)

**Comment by ccorcos** [comment]
@DPirate, here's a concrete example of the difference.

[playground link](https://www.typescriptlang.org/play?#code/FAehAIFkFMBcAsD2ATAzuADgQwE5YLZzQ7q7TgBGAlgG65VYB2sAdMAMYA2Wq6AGgB4AKgD5wAb2DhpmAK6p4ACiqxo+AFzghASgkBfYAeAAzWY3awqiRuAAeARkWcqqWJsGucVRgHNwAH3BGWXwKYhFdSWlnVxYMeSUAInhoTk5ERO1DYAdFRmgAd3APWC9fEUVtLLBwAANxeIVlVQ1wT28fbU0aRCpkPVrwF3AeVCofRiwKTnJYRDqGhOa1TXbfAKCQsJwu8B6+geATMwsrG1sAJicXN2KBNZ8IiSlwGNZGpJS0jKyDS7zCncHhtgqFwpVqhB6h9lq1gYFQdtdvt+oNhqNxpNprN5tClioVm1Sh1kb1UUdQBAAGInSzWTC4AhEEgjHDkdjWUpYOheJiwDjcXjgACawjEURkH3AAF5wLDNDoZeKDEZTOY6TYAJ6ON6aUXwzZgnBPCVvOJLZKpdKZbLagFFfXE8oQ8A1ISajDkACiOBwiBwlIWMIJcKdnW6ZIGQ3QADkAPJCEa8TFTGbgOZB-EtVZhkFbYikg61I5q070zVXXUi+65k0vM0wy3fG0GCv26sGxHgqqBvFNEM5sp+BH5nYRovRpNjCapnGZ-vZolDwvk4BAA)

```ts
// Methods parameters are bivariant.
class X<T> {
    push(item: T) {}
}

function x1(list: X<string | number>) {
  list.push("hello")
}
x1(new X<string>())
// `{push(item: string): void}` is assignable to `{push(item: string | number): void}`

function x2(list: X<string>) {
  list.push("hello")
}
x2(new X<string | number>())
// `{push(item: string | number): void}` is assignable to `{push(item: string): void}`


// Function parameters are contravariant
class Y<T> {
    push = (item: T) => {}
}

function y1(list: Y<string | number>) {
  list.push("hello")
}
y1(new Y<string>()) // Type Error
// `{push(item: string): void}` is NOT assignable to `{push(item: string | number): void}`

function y2(list: Y<string >) {
  list.push("hello")
}
y2(new Y<string | number>())
// `{push(item: string | number): void}` is assignable to `{push(item: string): void}`
```

In terms of _why_ is it designed this way: https://github.com/microsoft/TypeScript/wiki/FAQ#why-are-function-parameters-bivariant

**Comment by DPirate** [comment]
> ```ts
> // Function parameters are contravariant
> class Y<T> {
>     push = (item: T) => {}
> }
> ```

Thanks a lot. I did not know this won't be considered as a method, will look into that.
